dast/pretty_diff.py

```diff
@@ -0,0 +1,168 @@
+"""
+Logic for printing diffs output by DeepDiff
+"""
+import ast
+import re
+from dataclasses import dataclass
+from copy import deepcopy
+from typing import Optional, Tuple
+
+
+GREEN = '\033[92m'
+RED = '\033[91m'
+CYAN = '\033[96m'
+YELLOW = '\033[93m'
+END = '\033[0m'
+
+@dataclass
+class Delta:
+    """
+    Node used to represent a change. Added to
+    an AST and represented with arrows by `Unparser`
+    """
+    old: ast.AST
+    new: ast.AST
+
+class Unparser(ast._Unparser):  # type: ignore[name-defined]
+    """
+    Specialisation of Unparser visitor with the added
+    ability to use colour to display diffs
+    """
+    def traverse(self, node):
+        if isinstance(node, list):
+            for item in node:
+                self.traverse(item)
+        else:
+            colour = getattr(node, "colour", None)
+            if colour is not None:
+                self._source.append(colour)
+            ast.NodeVisitor.visit(self, node)
+            if colour is not None:
+                self._source.append(END)
+
+    def visit_Delta(self, node):  # pylint: disable=invalid-name
+        """
+        Add arrows to the source code to represent changes.
+        """
+        before_old = len("".join(self._source).splitlines())
+        self.traverse(node.old)
+        after_old = len("".join(self._source).splitlines())
+        if after_old == before_old:
+            self._source.append(YELLOW + "->" + END)
+        else:
+            self._source.append("\n" + YELLOW + "\N{Downwards Arrow}" * 3)
+        self.traverse(node.new)
+
+def split_path(path: str) -> Tuple[str, str, Optional[int]]:
+    """
+    Split an AST path into the parent and the sub-path
+    to the child. For paths to iterables also produce the index
+
+    >>> split_path("a.b.c")
+    ('a.b', 'c', None)
+    >>> split_path("a.b.c[5]")
+    ('a.b', 'c', 5)
+    """
+    parent, _, child = path.rpartition(".")
+    match = re.match(r"(.*)\[(\d)]$", child)
+    index: Optional[int]
+    if match:
+        prop = match.group(1)
+        index = int(match.group(2))
+    else:
+        prop = child
+        index = None
+
+    return parent, prop, index
+
+def print_diff(diff, now_path, then: ast.AST, now: ast.AST):
+    """
+    Print the diff output by DeepDiff in a human-readable way.
+    """
+    print(f"diff --dast {now_path}")
+    both = deepcopy(now)
+    all_changes = []
+    for change_type, changes in diff.items():
+        for path, change in changes.items():
+            description, is_added = describe_change(then, now, change_type, path)
+            parent_path, prop, index = split_path(path)
+            all_changes.append((index, parent_path, prop, change, description, is_added))
+
+    for (
+            index,
+            parent_path,
+            prop,
+            change,
+            description,
+            is_added) in sorted(all_changes, key=lambda x: x[0] if x[0] is not None else -1):
```

Comment:
why the ordering matters? this indices are potentially totally unrelated isn't? ah I guess if the object belongs to the same iterable you want to process it in order and this will enforce it